{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structural Variation Calling - Solutions\n",
    "No questions in this section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at Structural Variants in VCF\n",
    "\n",
    "## Exercises\n",
    "\n",
    "1. What does the CIPOS format tag indicate? __Confidence interval around POS for imprecise variants__\n",
    "\n",
    "2. What does the PE tag indicate? __Number of paired-end reads supporting the variant across all samples__\n",
    "\n",
    "3. What tag is used to describe an inversion event? __INV__\n",
    "\n",
    "4. What tag is used to describe a duplication event? __DUP__\n",
    "\n",
    "5. How many deletions were called in total? (**Hint:** DEL is the info field for a deletion. The -c option of the grep command can be used to return a count of matches.) __31, try__\n",
    "\n",
    "__`grep -c \"<DEL>\" ERR1015121.vcf`__\n",
    "\n",
    "6. What type of event is predicted at IV:437148? What is the length of the SV? How many paired-end reads and split-reads support this SV variant call? __Deletion -370 20 PE 21 split__\n",
    "\n",
    "__`grep \"437148\" ERR1015121.vcf`__\n",
    "\n",
    "7. What is the total number of SV calls predicted on the IV chromosome? __10, try__\n",
    "\n",
    "__`grep -c \"^IV\" ERR1015121.vcf `__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calling Structural Variants\n",
    "\n",
    "**Q:** mean=454.87 std=86.29\n",
    "\n",
    "## Breakdancer\n",
    "\n",
    "### Exercises\n",
    "\n",
    "`grep \"83065\" ERR1015121.breakdancer.out`\n",
    "\n",
    "1. Inversion\n",
    "\n",
    "2. -116, \n",
    "\n",
    "3. 42\n",
    "\n",
    "`grep \"258766\" ERR1015121.breakdancer.out`\n",
    "\n",
    "4. Deletion (7325, 99)\n",
    "\n",
    "5.  `grep DEL | awk OFS= breakdancer.dels.bed | awk '{print $1\"\\t\"$2\"\\t\"$5\"\\t\"$7\"\\t\"$9}' > breakdancer.dels.bed`\n",
    "\n",
    "## Inspecting SVs with IGV\n",
    "\n",
    "### Exercises\n",
    "\n",
    "1. Yes, a deletion (view as paired, sort by insert size, squish).\n",
    "\n",
    "2. There are very few reads mapping, the reads that are mapped are of low mapQ and it has a SV score = 99 \n",
    "\n",
    "3. Size estimate? ~7.5k\n",
    "\n",
    "Was the deletion at II:258766 also called by the other structural variant software and was the predicted size?\n",
    "\n",
    "5. Yes, SVTYPE=DEL, SVLEN=-7438\n",
    "\n",
    "6. DEL called by breakdancer (score=59). Not found by other caller Lumpy.\n",
    "\n",
    "7. Yes, 2 reads support (red).\n",
    "\n",
    "## Dysgu\n",
    "\n",
    "### Exercises\n",
    "\n",
    "1. What was the total number of SVs identified? How many PASS SVs were identified by Dysgu? Why did the rest of the SVs fail?\n",
    "bcftools view -H ERR1015069.vcf | wc -l\n",
    "30\n",
    "\n",
    "bcftools view -H -i 'FILTER=\"PASS\"' ERR1015069.vcf |wc -l\n",
    "8\n",
    "\n",
    "lowProb: ##FILTER=<ID=lowProb,Description=\"Probability below threshold set with --thresholds\">\n",
    "\n",
    "3. What type of SV event occurs at position IV:384221? What is the length of the SV event? What is the genotype quality?\n",
    "\n",
    "DEL = Deletion, SVLEN=328, GQ=62\n",
    "\n",
    "3. What type of SV event occurs at position XV:31115? What is the length of the SV event? What is the probability of the structural variant?\n",
    "\n",
    "INS = Insertion, SVLEN=63, PROB=0.816\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Calling Structural Variants from Long Reads \n",
    "\n",
    "### Align the reads with minimap and convert to bam\n",
    "\n",
    "`minimap2 -t 2 -x map-pb -a ../ref/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa YPS128.filtered_subreads.10x.fastq.gz | samtools view -b -o YPS128.filtered_subreads.10x.bam -`\n",
    "\n",
    "### Sort the bam\n",
    "`samtools sort -T temp -o YPS128.filtered_subreads.10x.sorted.bam YPS128.filtered_subreads.10x.bam`\n",
    "\n",
    "`samtools calmd -b YPS128.filtered_subreads.10x.sorted.bam ../ref/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa > YPS128.filtered_subreads.10x.sorted.calmd.bam`\n",
    "\n",
    "### Index the sorted bam\n",
    "`samtools index YPS128.filtered_subreads.10x.sorted.calmd.bam`\n",
    "\n",
    "### Call SVs with sniffles\n",
    "`sniffles --input YPS128.filtered_subreads.10x.sorted.calmd.bam --vcf YPS128.filtered_subreads.10x.vcf`\n",
    "\n",
    "### Exercises\n",
    "1. What sort of SV was called at on chromosome ‘XV’ at position 854272? __Deletion_\n",
    "\n",
    "2. What is the length of the SV? __344__\n",
    "\n",
    "3. How many reads are supporting the SV? __14 (SUPPORT tag)__\n",
    "\n",
    "4. What sort of SV was called at on chromosome 'XI' at position 74608? __Insertion_\n",
    "\n",
    "5. What is the length of the SV? __358__\n",
    "\n",
    "6. How many reads are supporting the SV? __15__\n",
    "\n",
    "7. How many inversions were called in the VCF? Note inversions are denoted by the type 'INV'. __6 in total, 5 passed__\n",
    "\n",
    "8. How many duplications were called in the VCF? Note duplications are denoted by the type 'DUP'. __2__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bedtools\n",
    "\n",
    "## Exercises\n",
    "\n",
    "1. How many SVs found in `ERR1015069.dels.vcf` overlap with a gene? (**Hint:** Use bedtools intersect command)\n",
    "__18, try (note the -u parameter is required to get the unique number of SVs)__\n",
    "\n",
    "`bedtools intersect -u -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3  | wc -l`\n",
    "\n",
    "2. How many SVs found in `ERR1015069.dels.vcf` do not overlap with a gene? (**Hint:** note the -v parameter to bedtools intersect)\n",
    "__9, try__\n",
    "\n",
    "`bedtools intersect -v -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3  | wc -l`\n",
    "\n",
    "3. How many SVs found in `ERR1015069.dels.vcf` overlap with a more strict definition of 50%?\n",
    "__14, try__\n",
    "\n",
    "`bedtools intersect -u -f 0.5 -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3  | wc -l`\n",
    "\n",
    "4. How many features does the deletion at VII:811446 overlap with? What type of genes? Note you will need to also use the -wb option in bedtools intersect.\n",
    "`bedtools intersect -wb -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3 | grep 811446`\n",
    "__4 features, all of them are protein coding genes (biotype=protein_coding)__\n",
    "\n",
    "\n",
    "5. How many features does the deletion at XII:650823 overlap with? What type of genes? Note you will need to also use the -wb option in bedtools intersect.\n",
    "`bedtools intersect -wb -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3 | grep 811446`\n",
    "__2 features, all of them are protein coding genes (biotype=protein_coding)__\n",
    "\n",
    "4. What is the closest gene to the structural variant at IV:384220 in `ERR1015069.dels.vcf`?\n",
    "__YDL037C, try__\n",
    "\n",
    "`bedtools closest -d -a ERR1015069.dels.vcf -b Saccharomyces_cerevisiae.R64-1-1.82.genes.gff3| grep IV | grep 384220` \n",
    "\n",
    "5. How many SVs overlap between the two files `ERR1015069.dels.vcf` and `ERR1015121.dels.vcf`?\n",
    "__27, try__\n",
    "\n",
    "`bedtools intersect -u -a ERR1015069.dels.vcf -b ERR1015121.dels.vcf | wc -l`\n",
    "\n",
    "6. How many SVs have a 90% reciprocal overlap between the two files `ERR1015069.dels.vcf` and `ERR1015121.dels.vcf` (**Hint:** first find the option for reciprocal overlap by typing: bedtools intersect -h)\n",
    "__24, try__\n",
    "\n",
    "`bedtools intersect -u -r -f 0.9 -a ERR1015069.dels.vcf -b ERR1015121.dels.vcf | wc -l`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
