<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Solutions – Genome Sequence Analysis Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../course_modules/Module6/module6_instructor_notes.html" rel="next">
<link href="../../course_modules/Module6/module6_exercises.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Genome Sequence Analysis Course</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../welcome.html"> 
<span class="menu-text">Learning paths</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-modules" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Modules</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-modules">    
        <li>
    <a class="dropdown-item" href="../../course_modules/Module1/module1.html">
 <span class="dropdown-text">File formats</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module2/module2.html">
 <span class="dropdown-text">Reads alignment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module3/module3.html">
 <span class="dropdown-text">Variant calling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module4/module4.html">
 <span class="dropdown-text">Structural Variation Calling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module5/module5.html">
 <span class="dropdown-text">Genome assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module6/module6.html">
 <span class="dropdown-text">RNA-seq</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Module7/module7.html">
 <span class="dropdown-text">CHiP-Seq</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../course_modules/Glossary/glossary.html">
 <span class="dropdown-text">Glossary of terms</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../VM.html"> 
<span class="menu-text">How to setup a VM</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../Train-the-trainer.html"> 
<span class="menu-text">Train-the-trainer resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../meet the team.html"> 
<span class="menu-text">Meet the team</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../course_modules/Module6/module6.html">RNA-Seq</a></li><li class="breadcrumb-item"><a href="../../course_modules/Module6/module6_solutions.html">Solutions</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false">
 <span class="menu-text">File formats</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module1/module1_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Read alignment</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module2/module2_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">Variant calling</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module3/module3_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Structural Variation Calling</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module4/module4_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">Genome Assembly</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module5/module5_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">RNA-Seq</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_solutions.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module6/module6_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false">
 <span class="menu-text">CHiP-Seq</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_manual.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_instructor_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructor notes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../course_modules/Module7/module7_platform_guidance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Platform guidance</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introducing-the-tutorial-dataset" id="toc-introducing-the-tutorial-dataset" class="nav-link active" data-scroll-target="#introducing-the-tutorial-dataset">Introducing the tutorial dataset</a></li>
  <li><a href="#mapping-rna-seq-reads-to-the-genome-using-hisat2" id="toc-mapping-rna-seq-reads-to-the-genome-using-hisat2" class="nav-link" data-scroll-target="#mapping-rna-seq-reads-to-the-genome-using-hisat2">Mapping RNA-Seq reads to the genome using HISAT2</a></li>
  <li><a href="#map-convert-sam-to-bam-sort-and-index-using-the-reads-from-the-mt2-sample." id="toc-map-convert-sam-to-bam-sort-and-index-using-the-reads-from-the-mt2-sample." class="nav-link" data-scroll-target="#map-convert-sam-to-bam-sort-and-index-using-the-reads-from-the-mt2-sample.">Map, convert (SAM to BAM), sort and index using the reads from the MT2 sample.</a></li>
  <li><a href="#visualising-transcriptomes-with-igv" id="toc-visualising-transcriptomes-with-igv" class="nav-link" data-scroll-target="#visualising-transcriptomes-with-igv">Visualising transcriptomes with IGV</a></li>
  <li><a href="#transcript-quantification-with-kallisto" id="toc-transcript-quantification-with-kallisto" class="nav-link" data-scroll-target="#transcript-quantification-with-kallisto">Transcript quantification with Kallisto</a></li>
  <li><a href="#identifying-differentially-expressed-genes-with-sleuth" id="toc-identifying-differentially-expressed-genes-with-sleuth" class="nav-link" data-scroll-target="#identifying-differentially-expressed-genes-with-sleuth">Identifying differentially expressed genes with Sleuth</a></li>
  <li><a href="#interpreting-the-results" id="toc-interpreting-the-results" class="nav-link" data-scroll-target="#interpreting-the-results">Interpreting the results</a></li>
  <li><a href="#normalisation" id="toc-normalisation" class="nav-link" data-scroll-target="#normalisation">Normalisation</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../course_modules/Module6/module6.html">RNA-Seq</a></li><li class="breadcrumb-item"><a href="../../course_modules/Module6/module6_solutions.html">Solutions</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Solutions</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>To run the commands alongside the answers you must first have run all of the tutorial commands for that section and generated the output files they reference.</p>
<section id="introducing-the-tutorial-dataset" class="level3">
<h3 class="anchored" data-anchor-id="introducing-the-tutorial-dataset">Introducing the tutorial dataset</h3>
<p><strong>Q1: Why is there more than one FASTQ file per sample?</strong></p>
<p>There are 2 FASTQ files for each sample e.g.&nbsp;MT1_1.fastq and MT1_2.fastq. This is because this was paired-end sequence data.</p>
<p><code>ls data/MT1*.fastq.gz</code></p>
<p><code>ls data/MT1*.fastq.gz | wc -l</code></p>
<p>With Illumina paired-end sequencing, you will have a fragment (hopefully longer than your reads) which is sequenced at both ends. This means that there will be a “left” read (sometimes known as r1) and its corresonding “right” read (sometimes known as r2). These are indicated by the /1 and/2 at the end of the read name in the **_1.fastq and _2**.fastq files respectively.</p>
<p>So, the left read header _@HS18_08296:8:1101:1352:48181#4_ in MT1_1.fastq has the /1 suffix:</p>
<p>@HS18_08296:8:1101:1352:48181#4/1</p>
<p>And, the corresponding right read header in MT1_2.fastq has the /2 suffix:</p>
<p>@HS18_08296:8:1101:1352:48181#4/2</p>
<p><strong>Q2: How many reads are there for MT1?</strong></p>
<p>There are 2.5 million reads for MT1.</p>
<p>Ideally, you should count the reads in both files. This is because sometimes we have singletons (reads without a mate) after preprocessing steps such as trimming.</p>
<p>Our reads look like:</p>
<p>@HS18_08296:8:1101:1352:48181#4/2</p>
<p>ATCCGCCNANTTTNNNNATATAATTANNNGNAANNAANNNNAATNACANNNATTNNNNTAGNANNNGNNAGTNNACAAGGNTNNNNNNNAAAGNN</p>
<p>+</p>
<p>9AABE8A!D!DFA!!!!EE@CFCD@B!!!D!F6!!EE!!!!EE4!F5E!!!BEA!!!!B@6!A!!!D!!FD’!!C+D@@*!B!!!!!!!B&gt;DE!</p>
<p>With the FASTQ format there are four lines per read. So, as long as our files are not truncated we can count the number of lines and divide them by four.</p>
<p><code>zless data/MT1_1.fastq.gz | wc -l</code></p>
<p>So, we can then divide this by four to give us 1.25 million. We will get the same for our r2 reads:</p>
<p><code>zless data/MT1_2.fastq.gz | wc -l</code></p>
<p>So, we have 1.25 million read pairs or 2.5 million (1.25 x 2) reads for our MT1 sample.</p>
<p>You may also have thought ”aha, I can use the **@** which is at the start of the header”…</p>
<p><code>zgrep -c '^@' data/MT1_1.fastq.gz</code></p>
<p><code>zgrep -c '^@' data/MT1_2.fastq.gz</code></p>
<p>But, wait, there are 1,343,714 left reads and 1,250,000 right reads…can that be right…no.</p>
<p>Take a closer look at the quality scores on the fourth line…they also contain **@**. This is because the quality scores are Phred+33 encoded. For more information on quality score encoding see our Data Formats and QC tutorial or go here.</p>
<p>Instead, we can use the earlier information about the left and right read suffixes: /1 and /2. This can be with two commands:</p>
<p><code>zgrep -c '/1$' data/MT1_1.fastq.gz</code></p>
<p><code>zgrep -c '/2$' data/MT1_2.fastq.gz</code></p>
<p>Or, with only one command:</p>
<p><code>zgrep -c '/[12]$' data/MT1*.fastq.gz</code></p>
<p>Note: Don’t forget to use the -c option for grep to count the occurences and $ to make sure you’re only looking for \1 or \2 at the end of the line.</p>
</section>
<section id="mapping-rna-seq-reads-to-the-genome-using-hisat2" class="level3">
<h3 class="anchored" data-anchor-id="mapping-rna-seq-reads-to-the-genome-using-hisat2">Mapping RNA-Seq reads to the genome using HISAT2</h3>
</section>
<section id="map-convert-sam-to-bam-sort-and-index-using-the-reads-from-the-mt2-sample." class="level3">
<h3 class="anchored" data-anchor-id="map-convert-sam-to-bam-sort-and-index-using-the-reads-from-the-mt2-sample.">Map, convert (SAM to BAM), sort and index using the reads from the MT2 sample.</h3>
<p>You can do this with several, individual steps:</p>
<p><code>hisat2 --max-intronlen 10000 -x data/PccAS_v3_hisat2.idx \-1 data/MT2_1.fastq.gz -2 data/MT2_2.fastq.gz -S data/MT2.sam</code></p>
<p><code>samtools view -b -o data/MT2.bam data/MT2.sam</code></p>
<p><code>samtools sort -o data/MT2_sorted.bam data/MT2.bam</code></p>
<p><code>samtools index data/MT2_sorted.bam</code></p>
<p>Or, you can do it in one step:</p>
<p><code>hisat2 --max-intronlen 10000 -x data/PccAS_v3_hisat2.idx -1 data/MT2_1.fastq.gz -2 data/MT2_2.fastq.gz | samtools view -b - | samtools sort -o data/MT2_sorted.bam - &amp;&amp; samtools index data/MT2_sorted.bam</code></p>
<p>Or, you could write a for loop to run the command above for all five of your samples:</p>
<pre><code>for r1 in data/\*\_1.fastq.gz

do
  sample=\$(basename \$r1)

  sample=\${sample/\_1.fastq.gz/}

  echo "Processing sample: "\$sample

  hisat2 --max-intronlen 10000 -x data/PccAS_v3_hisat2.idx -1 "data/\${sample}\_1.fastq.gz" -2 "data/\${sample}\_2.fastq.gz" | samtools view -b - | samtools sort -o "data/\${sample}\_sorted.bam" - &amp;&amp; samtools index "data/\${sample}\_sorted.bam"
done</code></pre>
<p>For more information on how this loop works, have a look at our Unix tutorial and Running commands on multiple samples.</p>
<p><strong>Q1: How many index files were generated when you ran hisat2-build?</strong></p>
<p>There are 8 HISAT2 index files for our reference genome.</p>
<p><code>ls data/*.ht2</code></p>
<p><code>ls data/*.ht2 | wc -l</code></p>
<p><strong>Q2: What was the overall alignment rate for each of the MT samples (MT1 and MT2) to the reference genome?</strong></p>
<pre><code>The overall alignment rate for MT1 was 94.16% and % for MT2.

1250000 reads; of these:

1250000 (100.00%) were paired; of these:

105654 (8.45%) aligned concordantly 0 times

329304 (26.34%) aligned concordantly exactly 1 time

815042 (65.20%) aligned concordantly \&gt;1 times

------------------------------------------------------------------------

105654 pairs aligned concordantly 0 times; of these:

1797 (1.70%) aligned discordantly 1 time

------------------------------------------------------------------------

103857 pairs aligned 0 times concordantly or discordantly; of these:

207714 mates make up the pairs; of these:

146078 (70.33%) aligned 0 times

19877 (9.57%) aligned exactly 1 time

41759 (20.10%) aligned \&gt;1 times

94.16% overall alignment rate

1250000 reads; of these:

1250000 (100.00%) were paired; of these:

139440 (11.16%) aligned concordantly 0 times

483493 (38.68%) aligned concordantly exactly 1 time

627067 (50.17%) aligned concordantly \&gt;1 times

------------------------------------------------------------------------

139440 pairs aligned concordantly 0 times; of these:

4967 (3.56%) aligned discordantly 1 time

------------------------------------------------------------------------

134473 pairs aligned 0 times concordantly or discordantly; of these:

268946 mates make up the pairs; of these:

207579 (77.18%) aligned 0 times

28834 (10.72%) aligned exactly 1 time

32533 (12.10%) aligned \&gt;1 times

91.70% overall alignment rate</code></pre>
<p><em>Note: If a read pair is concordantly aligned it means both reads in the pair align with the same chromosome/scaffold/contig, the reads are aligned in a proper orientation (typically —-&gt; &lt;—-) and that the reads have an appropriate insert size.</em></p>
<p><strong>Q3: How many MT1 and MT2 reads were not aligned to the reference genome?</strong></p>
<p>146,250 reads (5.85%) and 207,729 reads (8.31%) did not align to the reference genome for MT1 and MT2 respectively.</p>
<p>Here is a brief summary of what the HISAT2 summary tells us for our MT2 sample and how we can tell which of the summary lines gives us this information:</p>
<p>• We have 1,250,000 read pairs or 2,500,000 reads (2 x 1,250,000 pairs)</p>
<p><code>1250000 reads; of these:</code></p>
<p>• All of our reads (100%) are paired - i.e.&nbsp;no reads without their mate</p>
<p><code>1250000 (100.00%) were paired; of these:</code></p>
<p>• 1,144,346 pairs (95.24%) align concordantly one (26.34%) or more (65.20%) times</p>
<p><code>329304 (26.34%) aligned concordantly exactly 1 time</code></p>
<p><code>815042 (65.20%) aligned concordantly &gt;1 times</code></p>
<p>• 105,654 pairs (8.45%) or 211,308 reads (2 x 105,654) did not align concordantly anywhere</p>
<p>in the genome</p>
<p><code>105654 (8.45%) aligned concordantly 0 times</code></p>
<p>• Of those 105,654 pairs, 1,797 pairs align discordantly (1.70% of the 105,654 pairs)</p>
<p><code>105654 pairs aligned concordantly 0 times; of these:</code></p>
<p><code>1797 (1.70%) aligned discordantly 1 time</code></p>
<p>• This leave us with 103,857 pairs (105,654 - 1,797) where both reads in the pair do not align</p>
<p>to the genome (concordantly or discordantly)</p>
<p><code>103857 pairs aligned 0 times concordantly or discordantly; of these:</code></p>
<p>• Of those 207,714 reads (2 x 103,857) we have 61,636 reads (29.67%) which align to the</p>
<p>genome without their mate</p>
<p><code>207714 mates make up the pairs; of these:</code></p>
<p><code>...</code></p>
<p><code>19877 (9.57%) aligned exactly 1 time</code></p>
<p><code>41759 (20.10%) aligned &gt;1 times</code></p>
<p>• Leaving us with a 94.16% overall alignment rate</p>
<p><code>94.16% overall alignment rate</code></p>
<p>• That means 146,078 (5.84%) of the 2,500,000 reads (or 70.33% of the unaligned pairs) do not align anywhere in the genome</p>
<p><code>146078 (70.33%) aligned 0 times</code></p>
</section>
<section id="visualising-transcriptomes-with-igv" class="level3">
<h3 class="anchored" data-anchor-id="visualising-transcriptomes-with-igv">Visualising transcriptomes with IGV</h3>
<p><strong>Q1: How many CDS features are there in “PCHAS_1402500”?</strong></p>
<p>There are 8 CDS features in PCHAS_1402500. You can get this in several ways:</p>
<p>Count the number of exons/CDS features in the gene annotation.</p>
<p>Count the number of CDS features in the GFF file.</p>
<p>First, get all of the CDS features for PCHAS_1402500.</p>
<p><code>grep -E "CDS.*PCHAS_1402500" data/PccAS_v3.gff3</code></p>
<p>Then count the number of the CDS features for PCHAS_1402500.</p>
<p><code>grep -cE "CDS.*PCHAS_1402500" data/PccAS_v3.gff3</code></p>
<p><strong>Q2: Does the RNA-seq mapping agree with the gene model in blue?</strong></p>
<p>Yes. The peaks of the coverage tracks correspond to the annotated exon/CDS features.</p>
<p><strong>Q3: Do you think this gene is differentially expressed and is looking at the coverage plots alone a reliable way to assess differential expression?</strong></p>
<p>Possibly. But, you can’t tell differential expression by the counts alone as there may be differences in the sequencing depths of the samples.</p>
</section>
<section id="transcript-quantification-with-kallisto" class="level3">
<h3 class="anchored" data-anchor-id="transcript-quantification-with-kallisto">Transcript quantification with Kallisto</h3>
<p>Use <code>kallisto quant</code> four more times, for the MT2 sample and the three SBP samples.</p>
<p>You can run the individual <strong><code>kallisto quant</code></strong> commands as you did for MT1 for each of the remaining samples:</p>
<p><code>kallisto quant -i data/PccAS_v3_kallisto -o data/MT2 -b 100 data/MT2_1.fastq.gz data/MT2_2.fastq.gz</code></p>
<p><code>kallisto quant -i data/PccAS_v3_kallisto -o data/SBP1 -b 100 data/SBP1_1.fastq.gz data/SBP1_2.fastq.gz</code></p>
<p><code>kallisto quant -i data/PccAS_v3_kallisto -o data/SBP2 -b 100 data/SBP2_1.fastq.gz data/SBP2_2.fastq.gz</code></p>
<p><code>kallisto quant -i data/PccAS_v3_kallisto -o data/SBP3 -b 100 data/SBP3_1.fastq.gz data/SBP3_2.fastq.gz</code></p>
<p>Or, you can write a for loop which will run kallisto quant on all of the samples:</p>
<pre><code>for r1 in data/\*\_1.fastq.gz

do

  echo \$r1

  sample=\$(basename \$r1)

  sample=\${sample/\_1.fastq.gz/}

  echo "Quantifying transcripts for sample: "\$sample

  kallisto quant -i data/PccAS_v3_kallisto -o "data/\$sample" -b 100 \\

  "data/\${sample}\_1.fastq.gz" "data/\${sample}\_2.fastq.gz"

done</code></pre>
<p>For more information on how this loop works, have a look at our Unix tutorial and Running commands on multiple samples.</p>
<p><strong>Q1: What k-mer length was used to build the Kallisto index?</strong></p>
<p>A k-mer length of 31 was used.</p>
<p>Look at the output from kallisto index:</p>
<p>[build] k-mer length: 31</p>
<p>Or, look for the <strong>-k</strong> or <strong>–kmer-size</strong> option in the <code>kallisto index</code> usage:</p>
<p><code>kallisto index</code></p>
<p><strong>Q2: How many transcript sequences are there in PccAS_v3_transcripts.fa?</strong></p>
<p>There are 5177 transcript sequences.</p>
<p>Look at the output from <code>kallisto quant</code>:</p>
<p>[index] number of targets: 5,177</p>
<p>Or, look for n_targets in one of the run_info.json files:</p>
<p><code>cat data/MT1/run_info.json</code></p>
<p>Or, you can run a grep on the transcript FASTA file and count the number of header lines:</p>
<p><code>grep -c "&gt;" data/PccAS_v3_transcripts.fa</code></p>
<p><strong>Q3: What is the transcripts per million (TPM) value for PCHAS_1402500 in each of the samples?</strong></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Sample</th>
<th>Transcripts Per Million (TPM)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MT1</td>
<td>2342.23</td>
</tr>
<tr class="even">
<td>MT2</td>
<td>1354.42</td>
</tr>
<tr class="odd">
<td>SBP1</td>
<td>2295.24</td>
</tr>
<tr class="even">
<td>SBP2</td>
<td>3274.98</td>
</tr>
<tr class="odd">
<td>SBP3</td>
<td>2536.17</td>
</tr>
</tbody>
</table>
<p>You can look at each of the individual abundance files:</p>
<p><code>grep "^PCHAS_1402500" data/MT1/abundance.tsv</code></p>
<p><code>grep "^PCHAS_1402500" data/MT2/abundance.tsv</code></p>
<p><code>grep "^PCHAS_1402500" data/SBP1/abundance.tsv</code></p>
<p><code>grep "^PCHAS_1402500" data/SBP2/abundance.tsv</code></p>
<p><code>grep "^PCHAS_1402500" data/SBP3/abundance.tsv</code></p>
<p>Or you can use a recursive grep:</p>
<p><code>grep -r "^PCHAS_1402500" data/*.tsv</code></p>
<p>Or you can use a loop:</p>
<pre><code>for r1 in data/\*\_1.fastq.gz

do

  sample=\$(basename \$r1)

  sample=\${sample/\_1.fastq.gz/}

  echo \$sample

  grep PCHAS_1402500 "data/\${sample}/abundance.tsv"

done</code></pre>
<p>For more information on how this loop works, have a look at our Unix tutorial and Running commands on multiple samples.</p>
<p><strong>Q4: Do you think PCHAS_1402500 is differentially expressed?</strong></p>
<p>Probably not. We would need to run statistical tests to really be sure though.</p>
</section>
<section id="identifying-differentially-expressed-genes-with-sleuth" class="level3">
<h3 class="anchored" data-anchor-id="identifying-differentially-expressed-genes-with-sleuth">Identifying differentially expressed genes with Sleuth</h3>
<p><strong>Q1: Is our gene from earlier, PCHAS_1402500, significantly differentially expressed?</strong></p>
<p>No.</p>
<p>Look at the transcript table:</p>
<p>And the transcript view.</p>
<p>Although this gene looked like it was differentially expressed from the plots in IGV, our test did not show it to be so (q-value &gt; 0.05). This might be because some samples tended to have more reads, so based on raw read counts, genes generally look up-regulated in the SBP samples.</p>
<p>Alternatively, the reliability of only two biological replicates and the strength of the difference between the conditions was not sufficient to be statistically convincing. In the second case, increasing the number of biological replicates would give us more confidence about whether there really was a difference.</p>
<p>In this case, it was the lower number of reads mapping to MT samples that mislead us in the IGV view. Luckily, careful normalisation and appropriate use of statistics saved the day!</p>
<p><code>grep PCHAS_1402500 data/kallisto.results | cut -f1,4</code></p>
</section>
<section id="interpreting-the-results" class="level3">
<h3 class="anchored" data-anchor-id="interpreting-the-results">Interpreting the results</h3>
<p><strong>Q1: How many genes are more highly expressed in the SBP samples?</strong></p>
<p>127. We can use awk to filter our Kallisto/sleuth results and wc -l to count the number of lines</p>
<p>returned.</p>
<p><code>awk -F"\t" '$4 &lt; 0.01 &amp;&amp; $5 &gt; 0' data/kallisto.results | wc -l</code></p>
<p><strong>Q2: How many genes are more highly expressed in the MT samples?</strong></p>
<p>169. We can use awk to filter our Kallisto/sleuth results and wc -l to count the number of lines returned.</p>
<p><code>awk -F"\t" '$4 &lt; 0.01 &amp;&amp; $5 &lt; 0' data/kallisto.results | wc -l</code></p>
<p><strong>Q3: Do you notice any particular genes that come up in the analysis?</strong></p>
<p>Genes from the <strong><em>cir</em></strong> family are upregulated in the MT samples.</p>
<p>To get this we must first find out which genes (or gene descriptions) are seen most often in the genes</p>
<p>which are more highly expressed in our SBP samples.</p>
<p><code>awk -F"\t" '$4 &lt; 0.01 &amp;&amp; $5 &gt; 0 {print $2}' data/kallisto.results | sort | uniq -c | sort -n</code></p>
<p>Then, we summarise the genes more highly expressed in our MT samples.</p>
<p><code>awk -F"\t" '$4 &lt; 0.01 &amp;&amp; $5 &lt; 0 {print $2}' data/kallisto.results | sort | uniq -c | sort -n</code></p>
<p>Perhaps the CIR proteins are interesting. There are only 2 CIR proteins upregulated in the SBP samples and 25 CIR in the MT samples.</p>
<p>The <strong><em>cir</em></strong> family is a large, malaria-specific gene family which had previously been proposed to be involved in immune evasion (Lawton et al., 2012). Here, however, we see many of these genes upregulated in a form of the parasite which seems to cause the immune system to better control the parasite. This suggests that these genes interact with the immune system in a subtler way, preventing the immune system from damaging the host.</p>
</section>
<section id="normalisation" class="level3">
<h3 class="anchored" data-anchor-id="normalisation">Normalisation</h3>
<p><strong>How we got the information to help the questions</strong></p>
<p>To answer the questions you needed the following for each sample:</p>
<p>• Number of reads assigned to PCHAS_1402500</p>
<p>• Length of exons in PCHAS_1402500 (bp)</p>
<p>• Total number of reads mapping</p>
<p>• Total RPK</p>
<p>First, take a quick look at the first five lines of the abundance.tsv for MT1.</p>
<p><code>head -5 data/MT1/abundance.tsv</code></p>
<p>There are five columns which give us information about the transcript abundances for our MT1 sample.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<tbody>
<tr class="odd">
<td><strong>Column</strong></td>
<td>Description</td>
</tr>
<tr class="even">
<td>target_id</td>
<td>Unique transcript identifier</td>
</tr>
<tr class="odd">
<td>length</td>
<td>Number of bases found in exons.</td>
</tr>
<tr class="even">
<td>eff_length</td>
<td><strong>Effective length</strong>. Uses fragment length distribution to determine the effective number of positions that can be sampled on each transcript.</td>
</tr>
<tr class="odd">
<td>est_counts</td>
<td><strong>Estimated counts</strong>. This may not always be an integer as reads which map to multiple transcripts are fractionally assigned to each of the corresponding transcripts.</td>
</tr>
<tr class="even">
<td>tpm</td>
<td><strong>Transcripts per million</strong>. Normalised value accounting for length and sequence depth bias.</td>
</tr>
</tbody>
</table>
<p>First, look for your gene of interest, PCHAS_1402500. Run this as a loop to grep the information for all five samples.</p>
<pre><code>for r1 in data/\*\_1.fastq.gz

do

  sample=\$(basename \$r1)

  sample=\${sample/\_1.fastq.gz/}

  echo \$sample

  grep PCHAS_1402500 "data/\$sample/abundance.tsv"

done</code></pre>
<p>Now you have the length (eff_length) and counts (est_counts) for PCHAS_1402500 for each of your samples. Next, you need to get the total number of reads mapped to each of your samples. You can use a loop to do this.</p>
<p>In the loop below samtools flagstat gives you the number of mapped paired reads (reads with itself and mate mapped) and those where one read mapped but its mate didn’t (singletons). If then uses <code>grep</code> to get the relevant lines and <code>awk</code> to add the mapped paired and singleton read totals together.</p>
<pre><code>for r1 in data/*_1.fastq.gz
do
  sample=$(basename $r1)
  sample=\${sample/_1.fastq.gz/}
  total=` samtools flagstat "data/${sample}_sorted.bam" |
  grep 'singletons \|with itself and mate mapped' | awk 'BEGIN{ count=0} {count+=\$1} END{print count}'\`
  echo -e "$sample\t\$total"
done</code></pre>
<p>Finally, to calculate the TPM values, you need the total RPK for each of your samples. Again we use a loop. Notice the use of NR&gt;2 in the <code>awk</code> command which tells it to skip the two header lines at the start of the file. You will also notice that we divide the eff_length by 1,000 so that it’s in kilobases.</p>
<pre><code>for r1 in data/*_1.fastq.gz
do
  sample=$(basename $r1)

  sample=${sample/_1.fastq.gz/}

  awk -F"\t" -v sample="$sample" 'BEGIN{total_rpk=0;} NR\&gt;2 { rpk=$4/($3/1000); total_rpk+=rpk } END{print sample"\t"total_rpk}' "data/${sample}/abundance.tsv"
done</code></pre>
<p><strong>Q1: Using the abundance.tsv files generated by Kallisto and the information above, calculate the RPKM for PCHAS_1402500 in each of our five samples.</strong></p>
<table class="caption-top table">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th>Sample</th>
<th>Per million scaling factor</th>
<th>Reads per million (RPM)</th>
<th>Per kilobase scaling factor</th>
<th>RPKM</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MT1</td>
<td>2.353922</td>
<td>1079.450</td>
<td>3.697</td>
<td>292</td>
</tr>
<tr class="even">
<td>MT2</td>
<td>2.292421</td>
<td>1479.484</td>
<td>3.709</td>
<td>399</td>
</tr>
<tr class="odd">
<td>SBP1</td>
<td>2.329424</td>
<td>6269.662</td>
<td>3.699</td>
<td>1695</td>
</tr>
<tr class="even">
<td>SBP2</td>
<td>2.187862</td>
<td>7908.131</td>
<td>3.696</td>
<td>2140</td>
</tr>
<tr class="odd">
<td>SBP3</td>
<td>2.164148</td>
<td>6767.421</td>
<td>3.699</td>
<td>1830</td>
</tr>
</tbody>
</table>
<p><strong>Q2: Using the abundance.tsv files generated by Kallisto and the information above, calculate the TPM for PCHAS_1402500 in each of our five samples.</strong></p>
<table class="caption-top table">
<colgroup>
<col style="width: 12%">
<col style="width: 40%">
<col style="width: 38%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th>Sample</th>
<th>Per kilobase scaling factor</th>
<th>Reads per kilobase (RPK)</th>
<th>TPM</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MT1</td>
<td>3.697</td>
<td>687.30</td>
<td>2342</td>
</tr>
<tr class="even">
<td>MT2</td>
<td>3.709</td>
<td>914.50</td>
<td>1354</td>
</tr>
<tr class="odd">
<td>SBP1</td>
<td>3.699</td>
<td>3947.87</td>
<td>2295</td>
</tr>
<tr class="even">
<td>SBP2</td>
<td>3.696</td>
<td>4681.81</td>
<td>3275</td>
</tr>
<tr class="odd">
<td>SBP3</td>
<td>3.699</td>
<td>3959.78</td>
<td>2536</td>
</tr>
</tbody>
</table>
<p>Q3: Do these match the TPM values from Kallisto?</p>
<p><strong>Yes</strong>.</p>
<p>Well, almost. They may be a couple out because we rounded up to make the calculations easier.</p>
<p><strong>Q4: Do you think PCHAS_1402500 is differentially expressed between the MT and SBP samples?</strong></p>
<p><strong>Probably not.</strong></p>
<p>If we were to look at only the counts and RPKM values then it appears there is an 8 fold difference between the MT and SBP samples. However, when we look at the TPM values, they are much closer and so differential expression is less likely.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../course_modules/Module6/module6_exercises.html" class="pagination-link" aria-label="Exercises">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Exercises</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../course_modules/Module6/module6_instructor_notes.html" class="pagination-link" aria-label="Instructor notes">
        <span class="nav-page-text">Instructor notes</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>This work is licensed under a Creative Commons Attribution 4.0 International License. Reuse is encouraged with acknowledgement</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>